File: content/problems/1562-find-latest-group-of-size-m/solutions/boundary-length-union/Solution.java
class Solution {
    public int findLatestStep(int[] arr,int m){
        int[] length=new int[arr.length+2];
        int groups=0,answer=-1;
        for(int step=1;step<=arr.length;step++){
            int position=arr[step-1],left=length[position-1],right=length[position+1];
            if(left==m)groups--;
            if(right==m)groups--;
            int merged=left+right+1;
            length[position-left]=length[position+right]=merged;
            if(merged==m)groups++;
            if(groups>0)answer=step;
        }
        return answer;
    }
}

File: content/problems/1562-find-latest-group-of-size-m/solutions/boundary-length-union/Solution.kt
class Solution {
    fun findLatestStep(arr:IntArray,m:Int):Int{
        val length=IntArray(arr.size+2);
        var groups=0;
        var answer=-1;
        for(step in 1..arr.size){
            val position=arr[step-1];
            val left=length[position-1];
            val right=length[position+1];
            if(left==m)groups--;
            if(right==m)groups--;
            val merged=left+right+1;
            length[position-left]=merged;
            length[position+right]=merged;
            if(merged==m)groups++;
            if(groups>0)answer=step
        };
        return answer
    }
}
